### DISCUSSION TITLE
Proxy package repository that can autobuild universal2 wheels

### DISCUSSION THREAD

**Discussion by fluffy-critter** [devpi/devpi#1010] [Ideas]
Hi, lately I've been doing some desktop application development using Python, poetry, and pyInstaller. An issue that I've run into is that a lot of important packages (such as Pillow) are only built for macOS as separate `arm64` and `x86_64` wheels, and have no plans of also adding a `universal2` wheel to the mix. Pillow in particular suggests using `delocate-fuse` to manually merge the packages together whenever necessary (and have no interest in providing `universal2` builds because that option exists), which works well enough for one-off builds but that adds extra barriers to making automatic and repeatable builds.

What I'm wondering is what it would take to adapt a devpi instance to provide a proxy that can determine if a requested package is only available in an unfused form and then fetch the underlying components in order to fuse them together. I've poked around at the docs and looked at the plugin hooks but it's all very abstract and non-obvious to me, but I've also never tried running a package repository myself.

Is this a reasonable idea or is there a better way to accomplish my goal?

**Comment by fschulze** [comment]
I don't think you need to poke at devpi internals for that. Providing the fusing on demand seems complicated to me. I would do it as a build step for the desktop application or in CI:

First create an index having ``root/pypi`` as base and ``mirror_whitelist=*`` which will only provide the ``universal2`` wheels. The whitelist setting allows ``root/pypi`` releases to show up once packages of the same name are uploaded. Alternatively the whitelist can list all relevant packages explicitly.

As the build step: Check whether a new version of the required packages is available as ``arm64`` and ``x86_64``, but not ``universal2``. The ``application/json`` data for ``user/index/project[/version]`` should have the necessary info. Build the ``universal2`` package and upload it the the index.

After that the application can be built with the assumption that the ``universal2`` wheel is available.

**Comment by fluffy-critter** [comment]
That sounds promising from what I understand, although I'm still not particularly familiar with how package indices work (this is the first time I've had to delve down into the lower-level parts of how packages are actually served up from an index). Thanks!

Are there any API-type docs I can look at to see how one actually queries the index? I'm having difficulty wrapping my head around a lot of how these things work and I must not be forming the right search queries to find these things (every search query I try ends up pointing me to packages that happen to be hosted *on* pypi and fit the same keywords).

Also, is this something I could do entirely without having to spin up devpi, and using a local filesystem-based index? e.g. something I could configure in my `pyproject.toml` to look at the local directory first, before going out to pypi? That's obviously much more of a poetry question than a devpi question and it's totally fine if the answer is "ask the poetry folks instead." :)

Thanks!

**Comment by fschulze** [comment]
For understanding devpi indexes, maybe this helps: https://devpi.net/docs/devpi/devpi/stable/+d/userman/devpi_concepts.html

The benefit in using an index would be caching, as the package would be available from the index as an ``universal2`` wheel once it is uploaded. When doing it during build, it would be repeated each time.

One can do a lot from ``pyproject.toml`` with additional requirements for the build system. For devpi itself I wrote ``setuptools_changelog_shortener`` which is used for the readme: https://github.com/devpi/devpi/blob/main/common/pyproject.toml, but that is really in the realm on how Python package building works and has nothing to do with devpi.

**Reply by fluffy-critter** [reply]
Okay, thanks, I'll read up on that stuff!

Running it as a cache also totally makes sense in this context. Is there a pattern from which one can automate spawning/destroying a local devpi instance during project build?

And yeah I figured this is getting way outside of devpi's wheelhouse (no pun intended). I had no idea the build tools were so expressive, but it makes sense in retrospect.

**Comment by fluffy-critter** [comment]
Okay, between this thread and the conversation we had by email I think I have a workable solution. Thanks again for your help!
